tools/codegen/gen_functionalization_type.py

```diff
@@ -0,0 +1,344 @@
+        # If we eventually have any, the codegen will need extra logic to support it - for example,
+        # when we call the at::meta:: API, all tensor args need to be converted to the meta device.
+        assert not a.type.is_tensor_like()
+
+    non_self_bindings = [dispatcher.argument(a) for a in non_self_args]
+    non_self_value_ctypes = [cpp.argument_to_value_type(a) for a in non_self_args]
+    non_self_capture_exprs = translate(non_self_bindings, non_self_value_ctypes, method=False)
+    captured_view_args_str = ', '.join(f'{arg.name} = {val.expr}' for arg, val in zip(non_self_args, non_self_capture_exprs))
+
+    multi_output = is_multi_output(f.func.returns)
+
+    forward_lambda_decls = 'const at::Tensor& base, int64_t mutated_view_idx'
+    reverse_lambda_decls = 'const at::Tensor& base, const at::Tensor& mutated_view, int64_t mutated_view_idx'
+
+    if Variant.method in f.variants:
+        forward_lambda_call_str = f'base.{api_name}({", ".join(a.name for a in non_self_args)})'
+    else:
+        forward_lambda_call_str = f'at::{api_name}({", ".join(["base"] + [a.name for a in non_self_args])})'
+    reverse_lambda_args = ['base', 'mutated_view']
+
+    if is_multi_output(f.func.returns):
+        reverse_lambda_args.append('mutated_view_idx')
+        maybe_idx = '[mutated_view_idx]'
+    else:
+        maybe_idx = ''
+
+    reverse_lambda_args_str = ', '.join(reverse_lambda_args + [a.name for a in non_self_args])
+
+    meta_call_args = [f'{a.name}_.to(kMeta)' if a.type.is_tensor_like() else a.name for a in f.func.arguments.flat_all]
+    if Variant.method in f.variants:
+        meta_call_str = f'{meta_call_args[0]}.{api_name}({", ".join(meta_call_args[1:])})'
+    else:
+        meta_call_str = f'at::{api_name}({", ".join(meta_call_args)})'
+
+    view_meta_str = f"""
+      at::functionalization::ViewMeta view_meta = at::functionalization::ViewMeta(
+        [{captured_view_args_str}]({forward_lambda_decls}) -> at::Tensor {{
+          return {forward_lambda_call_str}{maybe_idx};
+        }},
+        [{captured_view_args_str}]({reverse_lambda_decls}) -> at::Tensor {{
+          return at::functionalization::impl::{api_name}_inverse({reverse_lambda_args_str});
+        }}
+      );"""
+
+    if f.tag is not None and f.tag is Tag.inplace_view:
+        # See Note [Functionalization Pass - Inplace View Ops] for more details
+        return f"""
+      {view_meta_str}
+      at::functionalization::impl::mutate_view_meta(self, view_meta);
+      {unwrap_tensor_args_str(f)}
+      {return_type} reference_tensor_output;
+      {{
+        at::AutoDispatchSkipFunctionalize guard;
+        reference_tensor_output = {meta_call_str};
+      }}
+      // See  Note [Propagating strides in the functionalization pass]
+      at::functionalization::impl::set_strides(self, reference_tensor_output);
+      return self;
+"""
+
+    else:
+        return f"""
+      {unwrap_tensor_args_str(f)}
```

Comment:
There's a ton of manual "getting the names to line up just right" logic going on here. Let me describe a less manual way of getting the same effect.

At a high level, the idea is to decompose the problem into two parts:

1. Describe *what is in scope* (via a list of Bindings). As you define extra bindings (as is the case in `unwrap_tensor_args_str`), you increase the number of values that are in scope.
2. Describe *what is needed* (also via a list of Bindings). Once you have this, you can use translate to get from stuff that's in scope to stuff that is needed.

Using Bindings to represent both of these situations can pay off quite a bit if you a function API that you are both calling and writing the body for; you use the bindings to construct the call sites, and then also use the bindings to setup the context. It's win win.

Here, there one thing that is slightly nontrivial: we must distinguish between the outer FunctionalTensor and the inner, unwrapped Tensor. This can probably done by mangling the name with the ctypes here, so you have one name for functional tensors and another name for the unwrapped tensors, and then an extra argument for sigs that let you control what mangling occurs.